go/ngql/types.go

```diff
@@ -0,0 +1,405 @@
+// returning one or more *noms* values whose presence is indicated by the provided arguments.
+type getSubvaluesFn func(v types.Value, args map[string]interface{}) (interface{}, error)
+
+// Note: Always returns a graphql.NonNull() as the outer type.
+func nomsTypeToGraphQLType(t *types.Type, tm typeMap) graphql.Type {
+	gqlType, ok := tm[t.Hash()]
+	if ok {
+		return gqlType
+	}
+
+	// In order to handle cycles, we eagerly create the type so we can put it into the cache before
+	// creating any subtypes. Since all noms-types are non-nullable, the graphql NonNull creates a
+	// handy piece of state for us to mutate once the subtype is fully created
+	newNonNull := &graphql.NonNull{}
+	tm[t.Hash()] = newNonNull
+
+	switch t.Kind() {
+	case types.NumberKind:
+		newNonNull.OfType = graphql.Float
+
+	case types.StringKind:
+		newNonNull.OfType = graphql.String
+
+	case types.BoolKind:
+		newNonNull.OfType = graphql.Boolean
+
+	case types.StructKind:
+		newNonNull.OfType = structToGQLObject(t, tm)
+
+	case types.ListKind, types.SetKind:
+		valueTyp := t.Desc.(types.CompoundDesc).ElemTypes[0]
+		newNonNull.OfType = graphql.NewList(nomsTypeToGraphQLType(valueTyp, tm))
+
+	case types.MapKind:
+		keyTyp := t.Desc.(types.CompoundDesc).ElemTypes[0]
+		valueTyp := t.Desc.(types.CompoundDesc).ElemTypes[1]
+		newNonNull.OfType = graphql.NewList(mapEntryToGraphQLObject(keyTyp, valueTyp, tm))
+
+	case types.RefKind:
+		newNonNull.OfType = refToGraphQLObject(t, tm)
+
+	case types.UnionKind:
+		newNonNull.OfType = unionToGQLUnion(t, tm)
+
+	case types.BlobKind, types.ValueKind, types.TypeKind:
+		panic(fmt.Sprintf("%d: type not impemented", t.Kind()))
+
+	case types.CycleKind:
+		panic("not reached") // we should never attempt to create a schedule for any unresolved cycle
+
+	default:
+		panic("not reached")
+	}
+
+	return newNonNull
+}
+
+// Creates a union of structs type.
+func unionToGQLUnion(typ *types.Type, tm typeMap) *graphql.Union {
+	unionTyps := typ.Desc.(types.CompoundDesc).ElemTypes
+	unionTypes := make([]*graphql.Object, len(unionTyps))
+
+	for i, unionTyp := range unionTyps {
+		if unionTyp.Kind() != types.StructKind {
+			panic("booh: grqphql-go only supports unions of structs")
+		}
+
+		unionType := nomsTypeToGraphQLType(unionTyp, tm).(*graphql.NonNull).OfType.(*graphql.Object)
+		unionTypes[i] = unionType
+	}
+
+	return graphql.NewUnion(graphql.UnionConfig{
+		Name:  getTypeName(typ),
+		Types: unionTypes,
+		ResolveType: func(p graphql.ResolveTypeParams) *graphql.Object {
+			tm := p.Context.Value(tmKey).(typeMap)
+			typ := p.Value.(types.Value).Type()
+			gqlType := tm[typ.Hash()].(*graphql.NonNull).OfType.(*graphql.Object)
+			return gqlType
+		},
+	})
+}
+
+func structToGQLObject(typ *types.Type, tm typeMap) *graphql.Object {
+	structDesc := typ.Desc.(types.StructDesc)
+	fields := graphql.Fields{}
+
+	structDesc.IterFields(func(name string, fieldTyp *types.Type) {
+		fieldType := nomsTypeToGraphQLType(fieldTyp, tm)
```

Comment:
I find it confusing that you have xxxTyp and xxxType. Maybe name this as gqlXxxType nomsXxxType?